common/src/Model/Brush.cpp

```diff
@@ -916,80 +946,100 @@ namespace TrenchBroom {
             BrushGeometry result;
-            
             for (const BrushVertex* vertex : m_geometry->vertices()) {
                 const Vec3& position = vertex->position();
                 if (vertexSet.count(position) == 0) {
-                    remaining.addPoint(position);
+                    moving.removeVertexByPosition(position);
                     result.addPoint(position);
                 } else {
-                    moving.addPoint(position);
                     result.addPoint(position + delta);
                 }
             }
             
-            assert(moving.vertexCount() == vertices.size());
             assert(remaining.vertexCount() + moving.vertexCount() == vertexCount());
             
+            // Special case, takes care of the first column.
+            if (moving.vertexCount() == vertexCount())
+                return acceptVertexMove(result);
+            
             // Will vertices be removed?
-            if (!allowVertexRemoval && result.vertexCount() < m_geometry->vertexCount())
-                return false;
+            if (!allowVertexRemoval) {
+                // All moving vertices must still be present in the result
+                for (const Vec3& movingVertex : moving.vertexPositions()) {
+                    if (!result.hasVertex(movingVertex + delta))
+                        return rejectVertexMove();
+                }
+            }
             
             // Will the result go out of world bounds?
             if (!worldBounds.contains(result.bounds()))
-                return false;
+                return rejectVertexMove();
             
             // Will the brush become invalid?
             if (!result.polyhedron())
-                return false;
+                return rejectVertexMove();
             
             // One of the remaining two ok cases?
             if ((moving.point() && remaining.polygon()) ||
                 (moving.edge() && remaining.edge()))
-                return true;
+                return acceptVertexMove(result);
             
             // Invert if necessary.
             if (remaining.point() || remaining.edge() || (remaining.polygon() && moving.polyhedron())) {
                 using std::swap;
                 swap(remaining, moving);
                 delta = -delta;
             }
-            
-            // Now check if any of the moving vertices would travel into or through the remaining fragment.
+
+            // Now check if any of the moving vertices would travel through the remaining fragment.
             for (const BrushVertex* vertex : moving.vertices()) {
                 const Vec3& oldPos = vertex->position();
                 const Vec3 newPos = oldPos + delta;
                 
+                // Skip moving vertices that end up inside the remaining fragment.
+                // These are handled in the allowVertexRemoval section above.
+                if (remaining.contains(newPos))
```

Comment:
I think this check is not necessary. If it is true, the first if condition inside the following loop is false for every face: Either the old position was above the face, or it was below, but then it will never be above if the vertex ended up inside the remaining fragment.

Not that it matters performance wise, but it may be misleading for the reader.